packages/core/saved-objects/core-saved-objects-api-server/src/apis/find.ts

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type {
+  SortOrder,
+  AggregationsAggregationContainer,
+  Id as EsId,
+} from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import type { SavedObject } from '@kbn/core-saved-objects-common';
+
+type KueryNode = any;
+
+/**
+ * @public
+ */
+export interface SavedObjectsFindOptionsReference {
+  type: string;
+  id: string;
+}
+
+/**
+ * @public
+ */
+export interface SavedObjectsPitParams {
+  id: string;
+  keepAlive?: string;
+}
+
+/**
+ *
+ * @public
+ */
+export interface SavedObjectsFindOptions {
+  type: string | string[];
+  page?: number;
+  perPage?: number;
+  sortField?: string;
+  sortOrder?: SortOrder;
+  /**
+   * An array of fields to include in the results
+   * @example
+   * SavedObjects.find({type: 'dashboard', fields: ['attributes.name', 'attributes.location']})
+   */
+  fields?: string[];
+  /** Search documents using the Elasticsearch Simple Query String syntax. See Elasticsearch Simple Query String `query` argument for more information */
+  search?: string;
+  /** The fields to perform the parsed query against. See Elasticsearch Simple Query String `fields` argument for more information */
+  searchFields?: string[];
+  /**
+   * Use the sort values from the previous page to retrieve the next page of results.
+   */
+  searchAfter?: EsId[];
+  /**
+   * The fields to perform the parsed query against. Unlike the `searchFields` argument, these are expected to be root fields and will not
+   * be modified. If used in conjunction with `searchFields`, both are concatenated together.
+   */
+  rootSearchFields?: string[];
+
+  /**
+   * Search for documents having a reference to the specified objects.
+   * Use `hasReferenceOperator` to specify the operator to use when searching for multiple references.
+   */
+  hasReference?: SavedObjectsFindOptionsReference | SavedObjectsFindOptionsReference[];
+  /**
+   * The operator to use when searching by multiple references using the `hasReference` option. Defaults to `OR`
+   */
+  hasReferenceOperator?: 'AND' | 'OR';
```

Comment:
Do we know how many times this is used? Seems like a potential for a very expensive search request to me.